File: 백준/Gold/22862. 가장 긴 짝수 연속한 부분 수열 （large）/가장 긴 짝수 연속한 부분 수열 （large）.py
import sys

input = sys.stdin.readline

N, K = map(int, input().split())
nums = list(map(int, input().split()))

end = 0
result = 0
cnt = 0 # 지운 횟수 count
erased = [False] * N

for start in range(N):
    while end < N:
        if nums[end] % 2 == 1 and cnt < K:
            erased[end] = True
            cnt += 1
            end += 1
            continue
        if nums[end] % 2 == 0:
            end += 1
            continue

        break

    result = max(result, end - start - cnt)

    if erased[start]:
        erased[start] = False
        cnt -= 1

print(result)

File: 백준/Gold/22862. 가장 긴 짝수 연속한 부분 수열 （large）/가장 긴 짝수 연속한 부분 수열 （large）.java

import java.io.*;
import java.util.*;


public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] line = br.readLine().split(" ");

        int n = Integer.parseInt(line[0]);
        int k = Integer.parseInt(line[1]);

        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int[] nums = new int[n];
        for(int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        boolean[] erased = new boolean[n];
        int end = 0;
        int result = 0;
        int cnt = 0;

        for(int start = 0; start < n; start++) {
            while(end < n) {
                if(nums[end] % 2 == 1 && cnt < k) {
                    erased[end] = true;
                    cnt++;
                    end++;
                    continue;
                }

                if(nums[end] % 2 == 0) {
                    end++;
                    continue;
                }

                break;
            }

            result = Math.max(result, end - start - cnt);

            if(erased[start]) {
                erased[start] = false;
                cnt -= 1;
            }
        }

        System.out.println(result);

    }
}
